File: src/main/java/io/github/binout/soccer/domain/season/SeasonService.java
package io.github.binout.soccer.domain.season;

import io.github.binout.soccer.domain.date.*;
import io.github.binout.soccer.domain.event.FriendlyMatchPlanned;
import io.github.binout.soccer.domain.event.LeagueMatchPlanned;
import io.github.binout.soccer.domain.player.Player;
import io.github.binout.soccer.domain.player.PlayerRepository;
import io.github.binout.soccer.domain.season.match.FriendlyMatch;
import io.github.binout.soccer.domain.season.match.LeagueMatch;
import io.github.binout.soccer.domain.season.match.Match;

import javax.enterprise.event.Event;
import javax.inject.Inject;
import java.time.LocalDate;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class SeasonService {

    @Inject
    PlayerRepository playerRepository;

    @Inject
    SeasonRepository seasonRepository;

    @Inject
    FriendlyMatchDateRepository friendlyMatchDateRepository;

    @Inject
    LeagueMatchDateRepository leagueMatchDateRepository;

    @Inject
    Event<FriendlyMatchPlanned> friendlyMatchPlannedEvent;

    @Inject
    Event<LeagueMatchPlanned> leagueMatchPlannedEvent;

    public LeagueMatch planLeagueMatch(Season season, LeagueMatchDate date) {
        TreeMap<Integer, List<Player>> treeMap = computeGamesPlayed(date, playerRepository.all().filter(Player::isPlayerLeague), leagueCounter(season));
        LeagueMatch leagueMatch = season.addLeagueMatch(date, extractPlayers(treeMap, LeagueMatch.MAX_PLAYERS, true));
        leagueMatchPlannedEvent.fire(new LeagueMatchPlanned(leagueMatch, getSubstitutes(season, leagueMatch)));
        return leagueMatch;
    }

    public void substitutePlayer(Season season, Match match, Player player) {
        Player by = getSubstitute(season, match);
        MatchDate matchDate = getMatchDate(match).orElseThrow(() -> new IllegalArgumentException("Unknown match date"));
        if (matchDate.isAbsent(by)) {
            throw new IllegalArgumentException(by.name() + " is not present for this date");
        }
        match.replacePlayer(player, by);
        matchDate.absent(player);
    }

    private Player getSubstitute(Season season, Match match) {
        List<Player> substitutes = getSubstitutes(season, match);
        if (substitutes.isEmpty()) {
            throw new IllegalArgumentException("No substitutes available");
        }
        return substitutes.iterator().next();
    }

    private Optional<? extends MatchDate> getMatchDate(Match match) {
        LocalDate date = match.date();
        if (match instanceof FriendlyMatch) {
            return friendlyMatchDateRepository.byDate(date.getYear(), date.getMonth(), date.getDayOfMonth());
        } else {
            return leagueMatchDateRepository.byDate(date.getYear(), date.getMonth(), date.getDayOfMonth());
        }

    }

    public FriendlyMatch planFriendlyMatch(Season season, FriendlyMatchDate date) {
        TreeMap<Integer, List<Player>> treeMap = computeGamesPlayed(date, playerRepository.all(), globalCounter(season));
        FriendlyMatch friendlyMatch = season.addFriendlyMatch(date, extractPlayers(treeMap, FriendlyMatch.MAX_PLAYERS, false));
        friendlyMatchPlannedEvent.fire(new FriendlyMatchPlanned(friendlyMatch, getSubstitutes(season, friendlyMatch)));
        return friendlyMatch;
    }

    private TreeMap<Integer, List<Player>> computeGamesPlayed(MatchDate date, Stream<Player> players, Function<Player, Integer> counter) {
        Map<Integer, List<Player>> playersByGamesPlayed = players.filter(date::isPresent).collect(Collectors.groupingBy(counter));
        return new TreeMap<>(playersByGamesPlayed);
    }

    private Set<Player> extractPlayers(TreeMap<Integer, List<Player>> treeMap, int maxPlayers, boolean goalPriority) {
        Set<Player> players = new HashSet<>();
        if (goalPriority) {
            treeMap.values().stream().flatMap(Collection::stream).filter(Player::isGoalkeeper).forEach(players::add);
        }
        Iterator<Map.Entry<Integer, List<Player>>> iterator = treeMap.entrySet().iterator();
        while (iterator.hasNext() && teamIsNotFull(players, maxPlayers)) {
            Map.Entry<Integer, List<Player>> entry = iterator.next();
            List<Player> playersForCurrentCount = entry.getValue();
            Collections.shuffle(playersForCurrentCount); // randomize list
            Iterator<Player> playerIterator = playersForCurrentCount.iterator();
            while (playerIterator.hasNext() && teamIsNotFull(players, maxPlayers)) {
                players.add(playerIterator.next());
            }
        }
        return players;
    }

    private boolean teamIsNotFull(Set<Player> players, int maxPlayers) {
        return players.size() < maxPlayers;
    }

    public List<Player> getSubstitutes(Season season, Match match) {
        SeasonStatistics statistics = season.statistics();
        MatchDate date = getMatchDate(match).orElseThrow(() -> new IllegalArgumentException("Unknown match date"));
        List<String> players = match.players().collect(Collectors.toList());
        Comparator<Player> gamesPlayedComparator = getPlayerComparator(season, match);
        return playerRepository.all()
                .filter(date::isPresent)
                .filter(p -> !players.contains(p.name()))
                .sorted(gamesPlayedComparator)
                .collect(Collectors.toList());
    }

    private Comparator<Player> getPlayerComparator(Season season, Match match) {
        Function<Player, Integer> counter = match instanceof LeagueMatch ? leagueCounter(season) : globalCounter(season);
        return (p1, p2) -> Integer.compare(counter.apply(p1), counter.apply(p2));
    }

    public List<FriendlyMatchDate> friendlyMatchDatesToPlan(Season season) {
        Set<LocalDate> dates = season.friendlyMatches().map(FriendlyMatch::date).collect(Collectors.toSet());
        return matchDatesToPlan(dates, friendlyMatchDateRepository.all());
    }

    public List<LeagueMatchDate> leagueMatchDatesToPlan(Season season) {
        Set<LocalDate> dates = season.leagueMatches().map(LeagueMatch::date).collect(Collectors.toSet());
        return matchDatesToPlan(dates, leagueMatchDateRepository.all());
    }

    private <T extends MatchDate> List<T> matchDatesToPlan(Set<LocalDate> dates, Stream<T> allDates) {
        LocalDate now = LocalDate.now();
        return allDates
                .filter(d -> d.date().isAfter(now) || d.date().isEqual(now))
                .filter(d -> !dates.contains(d.date()))
                .filter(MatchDate::canBePlanned)
                .collect(Collectors.toList());
    }

    private static Function<Player, Integer> leagueCounter(Season season) {
        return p -> season.statistics().leagueMatchPlayed(p);
    }

    private static Function<Player, Integer> globalCounter(Season season) {
        return p -> season.statistics().matchPlayed(p);
    }
}

File: src/main/kotlin/io/github/binout/soccer/application/SeasonService.kt
package io.github.binout.soccer.application

import io.github.binout.soccer.domain.date.FriendlyMatchDate
import io.github.binout.soccer.domain.date.FriendlyMatchDateRepository
import io.github.binout.soccer.domain.date.LeagueMatchDate
import io.github.binout.soccer.domain.date.LeagueMatchDateRepository
import io.github.binout.soccer.domain.player.Player
import io.github.binout.soccer.domain.player.PlayerName
import io.github.binout.soccer.domain.player.PlayerRepository
import io.github.binout.soccer.domain.season.*
import io.github.binout.soccer.domain.season.FriendlyMatch
import io.github.binout.soccer.domain.season.LeagueMatch
import java.time.LocalDate
import java.time.Month
import javax.enterprise.context.ApplicationScoped
import javax.enterprise.context.Initialized
import javax.enterprise.event.Observes
import javax.inject.Inject


@ApplicationScoped
class InitializeSeason {

    @Inject lateinit var seasonRepository: SeasonRepository

    fun initCurrentSeason(@Observes @Initialized(ApplicationScoped::class) init: Any) {
        val currentSeason = Season.currentSeasonName()
        val optSeason = seasonRepository.all().firstOrNull { s -> s.name == currentSeason }
        if (optSeason == null) {
            seasonRepository.replace(Season(currentSeason))
        }
    }
}


class AddFriendlyMatch(
        private val seasonRepository: SeasonRepository,
        private val friendlyMatchDateRepository: FriendlyMatchDateRepository,
        private val seasonPlanning: SeasonPlanning) {

    fun execute(seasonName: String, year: Int, month: Month, day: Int) {
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Can not replace match to season")
        val matchDate = friendlyMatchDateRepository.byDate(year, month, day) ?: throw IllegalArgumentException("Can not replace match to season")
        seasonPlanning.planFriendlyMatch(season, matchDate)
    }
}


class AddLeagueMatch(
        private val seasonRepository: SeasonRepository,
        private val leagueMatchDateRepository: LeagueMatchDateRepository,
        private val seasonPlanning: SeasonPlanning) {

    fun execute(seasonName: String, year: Int, month: Month, day: Int) {
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Can not replace match to season")
        val matchDate = leagueMatchDateRepository.byDate(year, month, day) ?: throw IllegalArgumentException("Can not replace match to season")
        seasonPlanning.planLeagueMatch(season, matchDate)
    }
}



class AddSeason(private val seasonRepository: SeasonRepository) {

    fun execute(seasonName: String) {
        if (seasonRepository.byName(seasonName) == null) {
            seasonRepository.replace(Season(seasonName))
        }
    }
}


class GetAllSeasons(private val seasonRepository: SeasonRepository) {

    fun execute(): List<Season> = seasonRepository.all()
}


class GetFriendlyMatches(
        private val seasonRepository: SeasonRepository,
        private val matchPlanning: MatchPlanning) {

    fun execute(seasonName: String): List<Pair<FriendlyMatch, List<Player>>> {
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Invalid season")
        return season.friendlyMatches().map { it to matchPlanning.getSubstitutes(season, it) }
    }
}


class GetLeagueMatches(
        private val seasonRepository: SeasonRepository,
        private val matchPlanning: MatchPlanning) {

    fun execute(seasonName: String): List<Pair<LeagueMatch, List<Player>>> {
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Invalid season")
        return season.leagueMatches().map { it to matchPlanning.getSubstitutes(season, it) }
    }
}



class GetNextFriendlyMatches(
        private val seasonRepository: SeasonRepository,
        private val matchPlanning: MatchPlanning) {

    fun execute(seasonName: String): List<Pair<FriendlyMatch, List<Player>>> {
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Invalid season")
        return season.friendlyMatches()
                .filter{ it.isNowOrFuture() }
                .map { it to matchPlanning.getSubstitutes(season, it) }
    }
}


class GetNextLeagueMatches(
        private val seasonRepository: SeasonRepository,
        private val matchPlanning: MatchPlanning) {

    fun execute(seasonName: String): List<Pair<LeagueMatch, List<Player>>> {
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Invalid season")
        return season.leagueMatches()
                .filter { it.isNowOrFuture() }
                .map { it to matchPlanning.getSubstitutes(season, it) }
    }
}



class GetSeason(private val seasonRepository: SeasonRepository) {

    fun execute(name: String): Season? = seasonRepository.byName(name)
}


class GetSeasonStats(private val seasonRepository: SeasonRepository) {

    fun execute(name: String): SeasonStatistics? = seasonRepository.byName(name)?.statistics()
}


class GetToPlanFriendlyMatches(private val seasonRepository: SeasonRepository,
                               private val seasonPlanning: SeasonPlanning) {

    fun execute(seasonName: String): List<FriendlyMatchDate> = seasonRepository.byName(seasonName)
            ?.let { s -> seasonPlanning.friendlyMatchDatesToPlan(s) }
            ?: throw IllegalArgumentException("Invalid season")
}


class GetToPlanLeagueMatches(private val seasonRepository: SeasonRepository,
                             private val seasonPlanning: SeasonPlanning) {

    fun execute(seasonName: String): List<LeagueMatchDate> = seasonRepository.byName(seasonName)
            ?.let { s -> seasonPlanning.leagueMatchDatesToPlan(s) }
            ?: throw IllegalArgumentException("Invalid season")
}


class SubstitutePlayerInFriendlyMatches(
        private val playerRepository: PlayerRepository,
        private val seasonRepository: SeasonRepository,
        private val matchPlanning: MatchPlanning) {

    fun execute(seasonName: String, date: LocalDate, name: String) {
        val playerName = PlayerName(name)
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Invalid season")
        val player = playerRepository.byName(playerName) ?: throw  IllegalArgumentException("Invalid player")
        val match = season.friendlyMatches().first { m -> m.date == date }
        matchPlanning.substitutePlayer(season, match, player)
    }
}


class SubstitutePlayerInLeagueMatches(
        private val playerRepository: PlayerRepository,
        private val seasonRepository: SeasonRepository,
        private val matchPlanning: MatchPlanning) {

    fun execute(seasonName: String, date: LocalDate, name: String) {
        val playerName = PlayerName(name)
        val season = seasonRepository.byName(seasonName) ?: throw IllegalArgumentException("Invalid season")
        val player = playerRepository.byName(playerName) ?: throw IllegalArgumentException("Invalid player")
        val match = season.leagueMatches().first { m -> m.date == date }
        matchPlanning.substitutePlayer(season, match, player)
    }
}